lib/windows_utils.pm

```diff
@@ -11,28 +11,65 @@ package windows_utils;
 
 use base Exporter;
 use Exporter;
-
 use strict;
 use warnings;
-
 use testapi;
 
-our @EXPORT = qw(wait_boot_windows);
+our @EXPORT = qw(select_windows_in_grub2 wait_boot_windows reboot_or_shutdown open_powershell_as_admin run_in_powershell use_search_feature);
 
-# makes sure splash screen appears and then boots to desktop
-# arguments: bootloader_time => seconds # now long to wait for splash to appear
-sub wait_boot_windows {
-    my %args            = @_;
-    my $bootloader_time = $args{bootloader_time} // 100;
+sub use_search_feature {
+    my ($string_to_search) = @_;
+    return unless ($string_to_search);
+
+    send_key_until_needlematch 'windows-search-bar', 'super-s';
+    wait_still_screen stilltime => 2, timeout => 15;
+    type_string "$string_to_search ", max_interval => 100, wait_still_screen => 0.5;
+}
 
-    assert_screen 'windows-screensaver', 600;
+sub select_windows_in_grub2 {
+    return unless (get_var('DUALBOOT'));
 
-    # Reset the consoles after the reboot: there is no user logged in anywhere
-    reset_consoles;
+    assert_screen "grub-reboot-windows", 125;
+    send_key "down" for (1 .. 2);
+    send_key "ret";
+}
+
+sub open_powershell_as_admin {
+    send_key_until_needlematch 'quick-features-menu', 'super-x';
+    wait_still_screen stilltime => 2, timeout => 15;
+    send_key_until_needlematch 'user-acount-ctl-allow-make-changes', 'shift-a';
+    assert_and_click 'user-acount-ctl-yes';
+    wait_still_screen stilltime => 2, timeout => 15;
+    assert_screen 'powershell-as-admin-window', 180;
+    assert_and_click 'window-max';
+}
+
+sub run_in_powershell {
+    my ($code) = @_;
+
+    type_string $code, max_interval => 125;
+    save_screenshot;
+    wait_screen_change(sub { send_key 'ret' }, 10);
+    assert_screen([qw(opened-explorer powershell-successful-operation powershell-enablefeature-no-restart powershell-ready-prompt)], 600);
```

Comment:
I had to come up with a way how to sync the code and VM state while using PowerShell. Normally we have a serial console, which is not the case in windows atm.

It seems like I would need to use 2 of them at the most. I will clean it up. Nice curiosity trigger! :)

Reply:
Update: 

1a. PowerShell successful operation in full screen 
1b. PowerShell ready prompt.
2. Opened Explorer 
3. Opened WSL image

I will remove the needle from point 1a. Since 1a and 1b deal with the same thing